### PR TITLE
More validation pt1

### PR DESCRIPTION
**Pull request by utas-tajona**
test coming soon

### CODE REVIEW

**Comment by ccmoloney** [comment]
@jonescc Would you be able to review and resolve the conflicts for this one?

**Comment by jonescc** [comment]
Sure can probably look at that today sometime

**Comment by jonescc** [comment]
Looks like there's a compile problem.

**Comment by utas-tajona** [comment]
survey optional silently merged

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureBetweenL5l95.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            return Validated.valid("Measure in l5/l95 range");
+        }
+
+        val keysStr = outOfRange.stream().map(Object::toString)
+                .reduce("", (acc, value) -> acc + "|" + value);
+        return Validated.invalid(
+                new StagedRowError(
+                        new ErrorID(target.getId(),
+                                target.getRef().getStagedJob().getId(),
+                                target.getRef().getSpecies() + " is superseeded"),
```
Comment (jonescc) [inline api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureBetweenL5l95.java:53]: Looks like the wrong error message?

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TooOldFutureDate.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        val now
+                = LocalDate.from(ZonedDateTime.now());
+       val earliest = LocalDate.parse(dateStr);
+      if ( target.getDate().isAfter(earliest) && target.getDate().isBefore(now)) {
+          return Validated.valid("date in range");
+      }
+        return Validated.invalid(new StagedRowError(
+                new ErrorID(target.getId(),
+                        target.getRef().getStagedJob().getId(),
+                        "Date must be between 01/01/1992 and Today"),
```
Comment (jonescc) [inline api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TooOldFutureDate.java:35]: message should include specified date from?

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TotalCheckSum.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import au.org.aodn.nrmn.restapi.model.db.enums.ValidationCategory;
+import au.org.aodn.nrmn.restapi.model.db.enums.ValidationLevel;
+import au.org.aodn.nrmn.restapi.validation.BaseFormattedValidator;
+import au.org.aodn.nrmn.restapi.validation.StagedRowFormatted;
+import cyclops.control.Validated;
+import lombok.val;
+
+import java.util.Map;
+
+public class TotalCheckSum extends BaseFormattedValidator {
```
Comment (jonescc) [inline api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TotalCheckSum.java:14]: This one isn't included for RLS or ATRC?

File: api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/DebrisZeroObsTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        formatted.setCode("dez");
+        formatted.setSpecies(ObservableItem.builder().observableItemName("Debris-Zero").build());
+        formatted.setTotal(0);
+        formatted.setInverts(0);
+        formatted.setMeasureJson(Collections.emptyMap());
+        val res = validationRule.valid(formatted);
+        assertTrue(res.isValid());
+    }
+
+    public void deBrisWith0ValueShouldbeOK() {
```
Comment (jonescc) [inline api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/DebrisZeroObsTest.java:31]: Name doesn't match test/expected result

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TooOldFutureDate.java (hunk 2 above)
Comment (utas-tajona) [reply api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TooOldFutureDate.java:35]: extra information not needed. the error  knows which  rows/column the error is linked to.

Comment (jonescc) [reply api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TooOldFutureDate.java:35]: yes that will include the date being validated but not the earliest date allowed which for RLS is 2006-01-01 not  01/01/1992

Comment (jonescc) [reply api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TooOldFutureDate.java:35]: Why pass it in and use it and then hard code the date in the error message

File: api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureBetweenL5l95Test.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                return null;
+            }
+
+            @Override
+            public Long getLmax() {
+                return null;
+            }
+        };
+        formatted.setSpeciesAttributesOpt(Optional.of(specAttribute));
+        formatted.setMeasureJson(ImmutableMap.<Integer, Integer>builder().put(1, 0).put(3, 1).put(4, 2).build());
```
Comment (jonescc) [inline api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureBetweenL5l95Test.java:68]: Do we need a test for in range here?

File: api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureUnderLmaxTest.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                return 20L;
+            }
+        };
+        formatted.setSpeciesAttributesOpt(Optional.of(specAttribute));
+        formatted.setMeasureJson(ImmutableMap.<Integer, Integer>builder().put(1, 0).put(3, 1).put(4, 2).build());
+        formatted.setMethod(1);
+        val validationRule = new MeasureUnderLmax();
+        val res = validationRule.valid(formatted);
+        assertTrue(res.isValid());
+    }
```
Comment (jonescc) [inline api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureUnderLmaxTest.java:63]: Do we need a test for out of range here?

File: api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/SpeciesInvertSizingTest.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                return null;
+            }
+        };
+        formatted.setSpeciesAttributesOpt(Optional.of(specAttribute));
+        formatted.setMethod(1);
+        formatted.setIsInvertSizing(false);
+        val validationRule = new SpeciesInvertSizing();
+        val res = validationRule.valid(formatted);
+        assertTrue(res.isValid());
+    }
```
Comment (jonescc) [inline api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/SpeciesInvertSizingTest.java:63]: should we test for failure here?

File: api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/SpeciesNotSuperseededTest.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+class SpeciesNotSuperseededTest extends FormattedTestProvider {
+
+    @Test
+    public void notSupperSeededShouldFSuccess() {
+        val formatted = getDefaultFormatted().build();
+        formatted.setIsInvertSizing(false);
+        val validationRule = new SpeciesNotSuperseeded();
+        val res = validationRule.valid(formatted);
+        assertTrue(res.isValid());
+    }
```
Comment (jonescc) [inline api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/SpeciesNotSuperseededTest.java:20]: Should we test for a failure here?

**Review by jonescc — CHANGES_REQUESTED** [review]
Looks good - some minor comments/changes required - we can probably add missing tests separately if required.

**Comment by jonescc** [comment]
Should note on the backlog item that we haven't done the Survey not done check as we are waiting for it to be specced.

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TooOldFutureDate.java (hunk 2 above)
Comment (utas-tajona) [reply api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TooOldFutureDate.java:35]: oh sorry i see what you mean, updating

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TotalCheckSum.java (hunk 3 above)
Comment (utas-tajona) [reply api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/TotalCheckSum.java:14]: I have added back the check as "Warning" but i had a discussion with Bene where this check was not usefull since total is macro generated. I will confirm tomorrow if i can remove it

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureBetweenL5l95.java (hunk 1 above)
Comment (jonescc) [reply api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureBetweenL5l95.java:53]: Error message should be something like 'Measure out of l5/l95 range' not 'species is superseeded'?

**Comment by jonescc** [comment]
Hmmm can't get any of the formatted tests to fail when testing in the app with Maria Island (after fixing other raw validation errors).  Looks like there's a null pointer exception at:

https://github.com/aodn/nrmn-application/pull/214/files#diff-e76ffe83bd03aeb09f31bc08ba5a3ab1df830eb608bf825a6b78adf0c50951f5R148

**Comment by jonescc** [comment]
> Looks like there's a null pointer exception at:
> 
> https://github.com/aodn/nrmn-application/pull/214/files#diff-e76ffe83bd03aeb09f31bc08ba5a3ab1df830eb608bf825a6b78adf0c50951f5R148

Yes species can be null for Survey not done.   Needs to cater for that.   I'll remove it for the moment and keep testing.

**Comment by jonescc** [comment]
After removing all the Survey not done rows, I now just sit there waiting for the validation to finish.   Need to speed up the validations somehow.   Its not usable at the moment.

I suspect the ui_species_attribute view with full data is way too slow.

**Comment by jonescc** [comment]
> I suspect the ui_species_attribute view with full data is way too slow.

Yes, looking in postgres (yes its still running) its running the following sql statement:

```
SELECT  observable_item_id as id,species_name, common_name, is_invert_sized, l5, l95, m
axabundance as max_abundance, lmax FROM  nrmn.ui_species_attributes   where observable_item_id = $1
```

**Comment by jonescc** [comment]
We're currently trying to calculate l5/l95/lmax data on the fly which requires processing all the existing data for each species record.

**Comment by jonescc** [comment]
Talking to @ccmoloney - we think we should materialise that view for the moment and look at how we refresh it or improve performance next iteration.

**Comment by jonescc** [comment]
Killing my validation job - its still running - so I can continue with other work.

File: api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureBetweenL5l95Test.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        formatted.setMethod(1);
+        val validationRule = new MeasureBetweenL5l95();
+        val res = validationRule.valid(formatted);
+        assertTrue(res.isInvalid());
```
Comment (jonescc) [inline api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureBetweenL5l95Test.java:72]: Not sure why this is invalid but may be reading the test incorrectly.   Isn't 3 and 4 between 2.5 and 10.2?

Comment (utas-tajona) [reply api/src/test/java/au/org/aodn/nrmn/restapi/validation/validators/formatted/MeasureBetweenL5l95Test.java:72]: they are keys not values

**Review by jonescc — COMMENTED** [review]
So now I can test in the application:

I get a javascript error when clicking on Checksum didn't match total error to select records where this was the case - error was Typeerror - cannot read property 'rowIndex' of undefined

**Comment by jonescc** [comment]
Also, I set superseded by on an observable item to another species name and do not get a validation error for that species.

**Comment by jonescc** [comment]
> Also, I set superseded by on an observable item to another species name and do not get a validation error for that species.

OK, I see why you commented out the test - but it doesn't look like this test uses ui_species_attributes so commenting out looks unnecessary?

**Comment by jonescc** [comment]
Still need to test the remaining checks - and need to leave now.  Can someone else pick this up?   @ccmoloney? Thanks

**Review by ccmoloney — APPROVED** [review]
Merging to allow testing by Bene
